feat: Onboarding tour (UI Tours)

maharshivpatel opened the pull request:
UI Tours like this can be created from Form Tour. ( it requires basic knowledge of how querySelector() works )

https://github.com/frappe/frappe/assets/39730881/806c5cb2-3927-4e76-8797-368255976f27

`no-docs` (internal feature for Frappe, assume breaking changes)

maharshivpatel committed:
feat: add fields for onboarding ui tours.

Fields required for UI Tours.

maharshivpatel committed:
feat: add onboarding and status on boot

Add List of all onboarding tours to frappe.boot and onboarding status on frappe.boot.user

maharshivpatel committed:
fix: typo in reference_doctype depends_on

maharshivpatel committed:
feat: form tour logic and ui tour routes

on client-side added some form logic required by ui tours and to generate page_route based on conditions. on server-side added logic to update Form Tour Settings.

maharshivpatel committed:
fix: clear dashboard_name only if list name is not Dashboard

maharshivpatel committed:
feat: onboarding class and tour on router change.

Added frappe.router on_change to find matching tours based on where user is and run them using frappe.ui.OnboardingTour.

maharshivpatel committed:
fix(minor): use matching next_tour if found

maharshivpatel committed:
feat: frappe framework basic UI tours.

maharshivpatel committed:
fix: set default for user onboarding_status

ankush commented on frappe/hooks.py:
```diff
@@ -30,6 +30,7 @@
 	"controls.bundle.js",
 	"report.bundle.js",
 	"telemetry.bundle.js",
+	"onboarding.bundle.js",
```
This should be loaded async if tour are pending. 

Since this is only required once, it doesn't make much sense to load it always?

maharshivpatel committed:
Merge branch 'develop' into onboarding-tour

maharshivpatel committed:
fix: linter for onboarding.bundle.js

ankush commented on frappe/public/js/onboarding.bundle.js:
```diff
@@ -0,0 +1,344 @@
+frappe.ui.OnboardingTour = class OnboardingTour {
```
Bundle files should ideally just import code from other files. This way code can be split and reused. 

Even if it's not required, let's move it to separate file for consistency as this is quite a huge file.

phot0n commented:
will the tour be given to each user in the system? or is it just for the first user?

(note: i haven't read the code)

edit: it seems to be user specific (seeing as there is a checkbox in user doctype) 

does the tour take into account the ui elements according to the role? like some users dont have access to the ssearch bar/awesomebar

ankush commented:
Some comments:

- For first onboarding there should be a dialog to begin/skip tour. 
- Make sure tours are skippable completely. (by reusing close or adding "skip")

ankush commented on frappe/core/doctype/form_tour_settings/form_tour_settings.py:
```diff
@@ -0,0 +1,15 @@
+		onboarding_tours = [[tour.form_tour, json.loads(tour.page_route)] for tour in self.form_tours]
+		frappe.db.set_single_value(
+			"Form Tour Settings", "onboarding_tours", json.dumps(onboarding_tours)
+		)
```
This is updating self, you should probably modify the `onboarding_tours` directly before save?

ankush commented on frappe/core/doctype/user/user.json:
```diff
@@ -92,6 +92,8 @@
   "generate_keys",
   "column_break_65",
   "api_secret",
+  "onboarding_tours_section",
```
it's hidden field so no section required.

maharshivpatel commented:
> will the tour be given to each user in the system? or is it just for the first user?
> 
> (note: i haven't read the code)
> 
> edit: it seems to be user specific (seeing as there is a checkbox in user doctype)
> 
> does the tour take into account the ui elements according to the role? like some users dont have access to the ssearch bar/awesomebar

If elements are not available in DOM it will skip the step.

shariquerik commented on frappe/boot.py:
```diff
@@ -68,6 +68,9 @@ def get_bootinfo():
 	bootinfo.home_folder = frappe.db.get_value("File", {"is_home_folder": 1})
 	bootinfo.navbar_settings = get_navbar_settings()
 	bootinfo.notification_settings = get_notification_settings()
+	bootinfo.onboarding_tours = frappe.parse_json(
+		frappe.db.get_single_value("Form Tour Settings", "onboarding_tours") or "[]"
+	)
```
We should not load this every time. If logged-in users have no remaining onboarding steps, we should return []. We should keep a flag all_onboarding_completed for each user.

maharshivpatel replied:
We would have to call twice one to check if all are completed and other to get tours if not. also making sure that all_tours_completed is updated on various occasions. right now it doesn't seem that big of a issue we can implement something if we have many more tours and face any issues?

shariquerik commented:
<img width="1348" alt="image" src="https://github.com/frappe/frappe/assets/30859809/942b550b-68da-431d-9d02-38153d5c20f2">

Looking at this it is not very clear. Add an info banner on top explaining what this settings page is used for with an example.

shariquerik commented on frappe/desk/doctype/form_tour/form_tour.js:
```diff
@@ -2,36 +2,64 @@
-				cur_frm.tour.init({ tour_name }).then(() => cur_frm.tour.start());
+			return {};
+		};
+		frm.fields_dict["reference_doctype"].get_query = function (doc) {
+			return {
+				filters: {
+					istable: 0,
+				},
+			};
+		};
```
```suggestion
		frm.set_query("reference_doctype", () => {
			return { filters: { istable: 0 } };
		});
```

shariquerik commented on frappe/desk/doctype/form_tour/form_tour.js:
```diff
@@ -2,36 +2,64 @@
+		frm.fields_dict["report_name"].get_query = function (doc) {
+			if (doc.reference_doctype) {
+				return {
+					filters: {
+						ref_doctype: doc.reference_doctype,
+					},
+				};
 			}
-			route_changed.then(() => {
-				const tour_name = frm.doc.name;
-				cur_frm.tour.init({ tour_name }).then(() => cur_frm.tour.start());
+			return {};
+		};
```
You should move this to reference_doctype method also use frm.set_query()

maharshivpatel replied:
I have changed it to frm.set_query() but move this to reference_doctype method will break in some cases because reference_doctype is used in multiple places so if it is set before list_name is set to report_name filters won't apply.

shariquerik commented on frappe/desk/doctype/form_tour/form_tour.js:
```diff
@@ -2,36 +2,64 @@
+				},
+			};
+		};
+		!frm.doc.ui_tour &&
```
```suggestion
		!frm.doc.ui_tour && !frm.is_new() &&
```

shariquerik commented on frappe/desk/doctype/form_tour/form_tour.js:
```diff
@@ -2,36 +2,64 @@
+			};
+		};
+		!frm.doc.ui_tour &&
+			frm.add_custom_button(__("Show Tour"), async () => {
+				const issingle = await check_if_single(frm.doc.reference_doctype);
+				let route_changed = null;
+
+				if (issingle) {
+					route_changed = frappe.set_route("Form", frm.doc.reference_doctype);
+				} else if (frm.doc.first_document) {
+					const name = await get_first_document(frm.doc.reference_doctype);
+					route_changed = frappe.set_route("Form", frm.doc.reference_doctype, name);
+				} else {
+					route_changed = frappe.set_route("Form", frm.doc.reference_doctype, "new");
+				}
+				route_changed.then(() => {
+					const tour_name = frm.doc.name;
+					cur_frm.tour.init({ tour_name }).then(() => cur_frm.tour.start());
+				});
```
Move this to a separate function

shariquerik commented on frappe/desk/doctype/form_tour/form_tour.js:
```diff
@@ -2,36 +2,64 @@
+		frm.set_value("reference_doctype", message?.ref_doctype || "");
+	},
+	async before_save(frm) {
+		if (
+			frm.doc.select_view == "List" &&
+			frm.doc.list_name == "Dashboard" &&
+			frm.doc.dashboard_name &&
+			frm.doc.reference_doctype
+		) {
+			frappe.throw(
+				"Referance Doctype and Dashboard Name both can't be used at the same time."
+			);
+		}
```
Avoid throwing error, clear & hide the other field if one field is set

maharshivpatel replied:
We don't know which one user intends to use  and there isn't always a relation between two. that's why i am trowing error instead to clearing.

shariquerik commented on frappe/desk/doctype/form_tour/form_tour.js:
```diff
@@ -125,3 +162,75 @@ async function get_first_document(doctype) {
+	let route = [frm.doc.view_name];
+	switch (route[0]) {
+		case "Workspaces":
+			frm.doc.list_name = "";
+			frm.doc.new_document_form = 0;
+			frm.doc.report_name = "";
+			frm.doc.page_name = "";
+			frm.doc.dashboard_name = "";
+			frm.doc.reference_doctype = "";
```
clearing fields you can handle this in the view_name method. just clear all fields if the view_name value is changed

maharshivpatel replied:
some cases like  `frm.doc.report_name = ""` depends on list_name as well. i can add this to respective field's method but this is less likely to fail as it will clear everything before saving. if you think that would still be better let me know!

shariquerik commented:
The user experience is not good if we check the reset tour checkbox and save it again gets unchecked.
Add a Reset Tour Button on Action Bar instead of a checkbox. Also call a separate method for the same in backend it will clean up before_save event

shariquerik commented on frappe/desk/doctype/form_tour/form_tour.py:
```diff
@@ -8,20 +8,54 @@
+					user_doc.save()
 
 	def on_update(self):
 		if frappe.conf.developer_mode and self.is_standard:
 			export_to_files([["Form Tour", self.name]], self.module)
```
This should be at the end

ankush commented:
@maharshivpatel mark comments are resolved if they're resolved from your end.

ankush commented on frappe/desk/doctype/form_tour_step/form_tour_step.json:
```diff
@@ -105,17 +121,103 @@
+   "fieldtype": "Data",
+   "label": "Element Selector",
+   "mandatory_depends_on": "eval:(doc.ui_tour)",
+   "reqd": 1
```
`reqd` is not required if it's not UI tour. You can remove it completely.

maharshivpatel committed:
fix: load onboarding_tours async

loads onboarding_tours if enable_onboarding is checked and form tours are available for user.

maharshivpatel committed:
fix: remove from settings from core module.

maharshivpatel committed:
fix: add form tour settings to desk module.

maharshivpatel committed:
fix: remove onboarding_tours_section

maharshivpatel committed:
fix: remove always required for element_selector.

maharshivpatel committed:
fix: minor changes as per review.

maharshivpatel committed:
fix: update module in json to  Desk

maharshivpatel committed:
Merge branch 'develop' into onboarding-tour

maharshivpatel committed:
fix: change onboarding_status type to long text

json type breaks unrelated tests and in mariadb json is alias for long text.

maharshivpatel committed:
Merge branch 'develop' into onboarding-tour

maharshivpatel committed:
Merge branch 'onboarding-tour' of https://github.com/maharshivpatel/frappe into onboarding-tour

maharshivpatel committed:
chore: fix linters

ankush commented:
@maharshivpatel 

- [x] patch to skip existing users is not present.
- [x] "close" still doesn't close/skip entire tour.  
- [x] bootinfo is cached so completed tours show up again
- [x] fixing copies. I'm moving tour to user doctype instead of ToDo. ToDo is sh!t TBH.

ankush committed:
fix: fieldtype json -> text

ankush committed:
chore: fix main workspace tour copy

- fix copy
- make it run on any workspace

ankush committed:
fix: clear boot cache on updating form tour

ankush committed:
fix: clear cached onboarding status on update

ankush committed:
refactor: replace todo tour with user tour

adding more users is most likely to improve adoption. so better place to
show tours is user doctype?

ankush committed:
fix: email notification

maharshivpatel committed:
fix: reload form_tour_settings

reload form_tour_settings and form_tour_settings_item before form_tour because form_tour on_update have get_doc for form_tour_settings.

maharshivpatel committed:
fix: change onboarding_status fieldtype

change onboarding_status to small text and handle null cases.

maharshivpatel committed:
fix: guess module if not set in form tour

if unable to guess save in Desk module.

maharshivpatel committed:
chore: fix linters

missing translate function and newline at end of file.

maharshivpatel committed:
Merge branch 'develop' into onboarding-tour

ankush commented on frappe/utils/install.py:
```diff
@@ -14,6 +14,8 @@ def before_install():
 	frappe.reload_doc("core", "doctype", "doctype_link")
 	frappe.reload_doc("desk", "doctype", "form_tour_step")
+	frappe.reload_doc("desk", "doctype", "form_tour_settings")
+	frappe.reload_doc("desk", "doctype", "form_tour_settings_item")
```
This will likely break existing sites. Let's just remove this for now, not important part of this PR.

ankush replied:
This is also the reason why patch test is failing.

ankush committed:
refactor: remove form tour settings

ankush committed:
fix: patch existing users/site

ankush commented on frappe/public/js/frappe/desk.js:
```diff
@@ -74,6 +74,23 @@ frappe.Application = class Application {
 
 		// page container
 		this.make_page_container();
+		if (
```
Is it possible to use something like `frappe.ready` here?

s-aga-r commented:
@maharshivpatel not a good experience for screens with 720p (ex. 1366 * 768).

![image](https://github.com/frappe/frappe/assets/63660334/98851c1e-7179-489b-8a79-88c54e15625e)
